sql.go

```diff
@@ -0,0 +1,45 @@
+package ipld_eth_statedb
+
+const (
+	GetContractCodePgStr = `SELECT data FROM public.blocks WHERE key = $1`
+	GetStateAccount      = `SELECT balance, nonce, code_hash, storage_root, removed FROM eth.state_cids
+						INNER JOIN eth.header_cids ON (
+							state_cids.header_id = header_cids.block_hash
+							AND state_cids.block_number = header_cids.block_number
```

Comment:
We use `block_number` here because it's indexed (and the hash is not) right?

Reply:
Sort of, both are indexed, `(block_hash, block_number)` tuple is the primary key for `eth.header_cids` so there is a compound btree index on that. You can search against the first member of a compound btree index with pretty much the same efficiency/speed as searching against a standalone btree index on that single column, but the performance degrades heavily with later members. So there is also a separate index on `block_number`.

It's unnecessary here because of the deterministic relationship between `block_hash` and `block_number` (the hash input includes the block number), but JOINing on a unique constraint like this (from the Postgres perspective) ensures we map to a single record.

But , in v4+ of ipld-eth-db every table has been denormalized by `block_number` for partitioning by TimescaleDB (`block_number` is the partition key). So providing `block_number`  as a constraint or argument for joins or queries becomes vital for speeding things up by helping the query planner figure out which partitions to route to.

Reply:
I just noticed on some of the more recent optimization work Thomas has [done](https://github.com/cerc-io/ipld-eth-db/blob/v4/db/migrations/00023_get_storage_at_functions.sql#L8), he JOINs only with the `block_hash`. Perhaps it is faster?